### DISCUSSION TITLE
适配 Twitter 推文页

### DISCUSSION THREAD

**Discussion by Kenshin** [Kenshin/simpread#2923] [站点适配]
> 适配了推特的推文页，效果如下：

![image](https://user-images.githubusercontent.com/81074/136520511-4099d465-3e6c-4214-8097-81f69ca38d5d.png)

![image](https://user-images.githubusercontent.com/81074/136521162-664a5dda-2cda-4ac5-958b-7ba897f28a92.png)

![image](https://user-images.githubusercontent.com/81074/136521373-3068883b-1f82-4022-a663-ff45ca88809b.png)

# 已适配

- [x] 推文（含有 `#` 或 `@` 会自动化转换为超链接
- [x] 包含链接预览
- [x] 包含转推
- [x] 包含图片（显示的为原图，非压缩的图片）
- [x] 推文串
- [x] Emoji
- [x] 推文的发布时间 
- [x] 推文标题 `[[name]]@[[username]] - [[status id]]`

# 未适配

- [ ] 视频
- [ ] Gif

# 特色

感谢 @cangyuyao 提供了新版的适配规则，详细说明 [请看这里](https://github.com/Kenshin/simpread/discussions/2923#discussioncomment-3458575)。

# 如何使用

1. 上述均在第三方适配源 https://github.com/Kenshin/simpread/discussions/2287 （内含导入方式）

2. 从黑名单去掉 `twitter.com`

![image](https://user-images.githubusercontent.com/81074/136522117-d1ae5c9c-a2f5-4312-8b5e-3184edc74c7d.png)

3. 将 `https://twitter.com/*/status/*` 加入到排除列表（ 因推文为动态加载，所以无法直接进入阅读模式）

![image](https://user-images.githubusercontent.com/81074/136522203-80698a0e-2a0f-47de-aa88-37f28e24a159.png)

# 为什么要适配这些

来自简悦的一个用户的建议，TA 经常通过推特获取资讯，故此我适配了上述几种类型的网站。

# 关联

- [适配 Thread Reader](https://github.com/Kenshin/simpread/discussions/2046)

- [适配 Typefully 内容页](https://github.com/Kenshin/simpread/discussions/3517)

- [适配即刻的文章页与转发页](https://github.com/Kenshin/simpread/discussions/2771)

- [适配了一组微博页面](https://github.com/Kenshin/simpread/discussions/3515)

**Comment by angel126** [comment]
![image](https://user-images.githubusercontent.com/8450915/149731494-2061d4e3-d081-4157-9ac7-d5a6c3f2e106.png)
我打开twitter 然后点击AA激活，就是这个页面不太对呀？

**Reply by Kenshin** [reply]
给我 URL 看下，另外应该是下图所示的方式（注意下图截图所示的适配规则）

e.g. https://twitter.com/wanglei001/status/1482927434337030144

![image](https://user-images.githubusercontent.com/81074/149733864-09260ff7-1d4d-45dc-8077-ff7ff1595e9c.png)

**Reply by angel126** [reply]
我试了你的这个URL 也是不行，或者是一直出现图4，配置文件已更新，刷新当期页面才能生效，刷新之后还是一样出现相同的提示。
我把第三方适配清除一次，重新再导入，还是一样的结果。
![image](https://user-images.githubusercontent.com/8450915/149746350-9780318f-4ee7-4d2e-9995-49570cc85e4d.png)
![image](https://user-images.githubusercontent.com/8450915/149746405-f539b3f8-2894-458e-b2e2-66bda7ed5d9d.png)
![image](https://user-images.githubusercontent.com/8450915/149746462-f090ae8d-0db8-4b93-9582-d8482d605702.png)
![image](https://user-images.githubusercontent.com/8450915/149747059-6775ef61-8212-430c-bff6-64892ba30347.png)

**Reply by Kenshin** [reply]
> 刷新当期页面才能生效，刷新之后还是一样出现相同的提示

这个问题是你多开了相同的 URL 造成的。

重启浏览器，然后只开选项页导入第三方适配源，然后重启浏览器

**Reply by Kenshin** [reply]
@angel126 

请在这里讨论，不要新开。

另外，是否按照下面的截图配置了？

![image](https://user-images.githubusercontent.com/81074/149871361-42ecd638-c1ce-406d-8c13-238a59ece6e2.png)

**Comment by cangyuyao** [comment]
示例推文一：https://twitter.com/FNN_News/status/1551400209342930944
![SNAG- 2022-7-25 23 27 25](https://user-images.githubusercontent.com/3854729/180802332-5d9a1955-47ed-4644-9b8c-fae15e59c3e7.jpg)

启动阅读模式后部分内容缺失：第二个tag #どうぞ 后；第三个tag #育児 后，我的界面是英文，因此末尾应为Translate Tweet：
![SNAG- 2022-7-25 23 29 39](https://user-images.githubusercontent.com/3854729/180802844-dbbff9f0-8990-4949-b08b-10dae269eda4.jpg)

---

示例推文二：https://twitter.com/Vg_akira/status/997642761728806913

第一次打开阅读模式，内容实际无异常，但报错正文获取异常（示例一没有这个现象）：
![SNAG- 2022-7-25 23 36 37](https://user-images.githubusercontent.com/3854729/180803206-15bcc252-8411-422a-ad18-cf2d4b1c7a39.jpg)

点击“是”之后获取了原文+整个页面上各种按钮+下面一串推荐推文：
![SNAG- 2022-7-25 23 41 58](https://user-images.githubusercontent.com/3854729/180804511-1dd9f9b3-225a-4142-9090-46711a691247.jpg)

**Reply by Kenshin** [reply]
> 示例一

确实有问题，你在试试其他页面是否也存在问题？

> 示例二

不用理这个提示，只是当正文字数较少时的提示。应该可以获取成功。

**Reply by cangyuyao** [reply]
> > 示例一
> 
> 确实有问题，你在试试其他页面是否也存在问题？

测试了一下，应该是因为tag后面没有空格而是直接接了标点符号（标点不会被判定为tag）。

Translate Tweet的Translate消失是因为跟末尾的链接识别到一起去了（tag、@和外链，只要是链接都有这个现象）
![image](https://user-images.githubusercontent.com/3854729/181776523-3642dc8d-d867-4f60-8ef0-ae49058b8ae8.png)

另外还有两个反馈：

1. 能否支持emoji？
2. 能否支持发推时间？

**Reply by Kenshin** [reply]
okay 我研究下。

推的正文 HTML 结构「很恶心」，一条正文套了 nnnn 个结构。

**Reply by cangyuyao** [reply]
发现个新问题，图片不是原图，推原图链接带&name=orig或者:orig，自己改了一下带前者的链接不能识别上传？后者可以上传。

**Reply by Kenshin** [reply]
我随便找个推文 e.g. https://twitter.com/asusmt/status/1554720298519633920 然后得到了 https://pbs.twimg.com/media/FZN53SPagAAps1d?format=png 应该是原图

**Reply by cangyuyao** [reply]
推图片尺寸最大会有4000+px。
https://pbs.twimg.com/media/FYZCs3VWQAARApe?format=jpg 尺寸 1199 × 795 px
https://pbs.twimg.com/media/FYZCs3VWQAARApe?format=jpg&name=orig 尺寸 1765 × 1170 px

我本身装了替换原图的userscript，昨天试过简悦可以和userscript一起用，就改了改直接解决emoji和原图了。
但正文发布时间竟然就一个啥都没有的span tag，没辙……

**Reply by Kenshin** [reply]
如果每个图片都能加 `&name=orig` 的话，这不是什么问题，能确认这个事情吗？

**Reply by cangyuyao** [reply]
可以的，原图脚本我用了很多年了

**Reply by Kenshin** [reply]
okay 我记录下。

> 就改了改直接解决emoji和原图了。

是什么意思？

**Reply by cangyuyao** [reply]
原图脚本+另外写了段替换掉emoji图片。因为简悦是提取网页上已有内容所以直接用脚本修改了，也是一样作用。

**Reply by Kenshin** [reply]
替换掉 Emoji 图片 Emoji 本身？还是直接去掉了？

**Reply by cangyuyao** [reply]
用图片的alt文本（unicode emoji）替换了图片

**Reply by Kenshin** [reply]
okay 我研究下。

**Reply by Kenshin** [reply]
但是 https://github.com/Kenshin/simpread/discussions/2923#discussioncomment-3223223 提到的问题仍没办法通过你说的方式解决吧？（我近期开始优化这里）

**Reply by cangyuyao** [reply]
是的，tag后标点的问题还是有，不是很常见所以我暂时没理

**Reply by Kenshin** [reply]
> 是的，tag后标点的问题还是有，不是很常见所以我暂时没理

okay 我看下是否复杂？如果有共性并且不是很复杂的话，我会修复下。

另外，推文还有什么需要优化地方？

**Reply by cangyuyao** [reply]
还有个个人问题：英数和中文之间的空格是排版优化自动加上的吗？感觉放在推文里有点奇怪……

**Reply by cangyuyao** [reply]
对了，标题还是用本来的<title>比较好，否则同一个账号的不同推文使用同一个标题，自动导入增强会识别到错误的页面。

**Reply by Kenshin** [reply]
中英文空格是简悦阅读模式的特性。
标题的话，推文并没有标题的概念，所以 `<title>` 最好的方案就是目前这个。
但确实存在重复问题，我一般先用 Live Editor 先修改标题。

**Reply by Kenshin** [reply]
Emoji 和 原图都 OK 了，日期的话，你觉得加到哪个位置笔记好？另外，如果是 Thread 的话，每个都加吗？（可能有点麻烦）

**Reply by Kenshin** [reply]
另外 https://twitter.com/FNN_News/status/1551400209342930944 问题调查清楚了。因为使用了 CJK 字符，不太好判断，先这样吧。 😂

**Reply by cangyuyao** [reply]
能不能新增一个发布日期的 property？
我觉得只用提取当前推文的日期就行了。

**Reply by cangyuyao** [reply]
> 另外 https://twitter.com/FNN_News/status/1551400209342930944 问题调查清楚了。因为使用了 CJK 字符，不太好判断，先这样吧。 😂

我看了下tag后接英文标点也不对啊……比如这条https://twitter.com/LoveGG8_SONE/status/1555390080188964864
总而言之tag后没打空格的都不能正确识别，只是CJK文法里没有随手空格的习惯所以出现得比较多而已。

**Reply by Kenshin** [reply]
> 我看了下tag后接英文标点也不对啊……比如这条https://twitter.com/LoveGG8_SONE/status/1555390080188964864
总而言之tag后没打空格的都不能正确识别，只是CJK文法里没有随手空格的习惯所以出现得比较多而已。

是的，Hashtag 必须要有个空格，否则我这边不好判断。

***

日期我可以加，但我问的是「**在页面什么位置加比较好**」

**Reply by cangyuyao** [reply]
那就开头第一行吧？

**Reply by Kenshin** [reply]
可以了，包括：原图、Emoji 转换和推文发布时间（不包含推文串）

**Reply by cangyuyao** [reply]
> > 我看了下tag后接英文标点也不对啊……比如这条https://twitter.com/LoveGG8_SONE/status/1555390080188964864
> > 总而言之tag后没打空格的都不能正确识别，只是CJK文法里没有随手空格的习惯所以出现得比较多而已。
> 
> 是的，Hashtag 必须要有个空格，否则我这边不好判断。

https://cdn.jsdelivr.net/gh/twitter/twitter-text@v1.13.4/js/twitter-text.js
抄了这里的 validHashtag 和 validMentionOrList 正则可以解决 # 和 @ 的问题。url我暂时还没遇到很大问题所以先不管了。

然后又发现了个新问题：这样的内容 `([@suguosuguo](https://twitter.com/suguosuguo))` 导入notion不会识别为链接？是因为有 @ 吗？

**Reply by cangyuyao** [reply]
新问题2：

回复的推文自动检测到的不是正文：

https://twitter.com/moth0127/status/1549290807358029824

![image](https://user-images.githubusercontent.com/3854729/183240596-89df6459-320b-477e-9cc9-4011a6c3e1b1.png)

**Reply by cangyuyao** [reply]
Hi，更新了推文适配规则：

## 修复

- `#hashtag` `@username` `URL` 的解析错误

- 无法提取本人回复

- 部分情况下当前推文被提取两次

## 新增

- 标题改为 `[[name]]@[[username]] - [[status id]]`，看了一下其他本地备份工具，带推文id的格式比较常见

- 分割线 Markdown 语法 `---`（原先只有 `<sr-hr>` 标签）

- 被转推文增加 `<blockquote>` 标签（原先只有 `<sr-twitter-retweet>` 标签）

## 以下问题没有解决

- 长链接后多余省略号

- 有 Reply to @username这样标记的回复推文只提取到该行，无法提取到正文

- Notion 对 `[@username](url)` 的识别失败

- Markdown 标记符号的转义（例如 `*^_`，会影响阅读/聚焦模式显示，有需要可以自行修改）

- Thread 中其他人的回复（有需要可以自行修改）

- 部分情况下“翻译推文”按钮引发的解析错误

**Reply by Kenshin** [reply]
Cool 

我测试了几个页面，发现应该是完美适配的，等我更新到官方适配规则，然后发布它。

**Comment by JJcdomat** [comment]
无法解析twitter页面。新安装插件，并开启无痕模式，按步骤123设置。最后还是无法正常显示。
(1、可在站点管理器看到`https://twitter.com/*/status/*` ，2、默认不在黑名单，3、手动加入排除列表。）
![image](https://github.com/Kenshin/simpread/assets/17878533/3e9cc2cb-6335-4121-a01e-d48ff11ce866)

**Reply by Kenshin** [reply]
okay 已确定，可能是推的页面规则又改变造成的...

适配后我会在这里通知你。

**Reply by Qusunyee** [reply]
无法适配

**Comment by whitewatercn** [comment]
```
"name"    : "twitter.com",
        "url"     : "https://twitter.com/*/status/*",
```        
twitter现已改名x.com，第三方适配源里的相关代码已无法使用，经测试，这里改成如下即可
```
"name"    : "x.com",
        "url"     : "https://x.com/*/status/*",
```

> https://gist.github.com/Kenshin/aa3188b22f2d5dcd130610ce41cf6a7b?permalink_comment_id=5831414#gistcomment-5831414


但是由于gist无法申请pr，请kenshin手动修改相关gist，或者新建一个gist？或者新建一个仓库专门放相关第三方适配源，方便pr管理？

> https://github.com/Kenshin/simpread/discussions/2287#discussioncomment-14793298


@Kenshin
